File: hailXOR.cpp
#include <bits/stdc++.h>

// using namespace std;

int highestPowerof2(int n) 
{ 
   int p = (int)std::log2(n); 
   return (int)std::pow(2, p);  
} 

int main(){
	int t{};
	std::cin >> t;
	while (t--){
		int x{};
		int n{};
		std::cin >> n >> x;
		int a[n] {};
		for(int i=0; i<n; i++){
			std::cin >> a[i];
		}

		int z{1}, i{0};
		while ((z <= x) && (i < n)){
			if (a[i] == 0){
				i++;
				continue;
			}
			int p {highestPowerof2(a[i])};
			for (int j=i+1; j<n; j++){
				if (a[j]^p < a[j]){
					a[j] ^= p;
					a[i] ^= p;
					z++;
					break;
				}
				else if (j == n-1){
					a[j] ^= p;
					a[i] ^= p;
					z++;
					break;
				}
			}
			if (i == n-1)
				break;
		}
		int r{0};
		r = x - z;
		if (r && n==2){
			if (r&1){
				a[0] = 1;
				a[1] ^= 1;
			}

		}
	for(int i=0; i<n; i++){
		std::cout << a[i] << " ";
	}
	std::cout << '\n';
	}
}

File: hailxor.py
import math 
  
def highestPowerof2(n): 
  
    p = int(math.log(n, 2)) 
    return int(pow(2, p))

for _ in range(int(input())):
	n, x   = map(int, input().split())
	a = list(map(int, input().split()))

	c = 0
	y = 0
	for i in a:
		c += bin(i).count('1')
		y ^= i
	# print(y, c)
	c -= bin(a[-1]).count('1')
	if x > c:
		if n == 2 and (x-c)&1:
			print(1, y^1)
		else:
			print("0 "*(n-1), y, sep="")
	else:
		z = 0
		i = 0

		while (z < x) and (i < n):
			if a[i] == 0:
				i += 1
				continue
			p = highestPowerof2(a[i])
			for j in range(i+1, n):
				if a[j]^p < a[j]:
					a[j] ^= p
					a[i] ^= p
					z += 1
					break
				elif j == n-1:
					a[j] ^= p
					a[i] ^= p
					z += 1
					break
			if i == n-1:
				break

		r = x - z
		if r and n == 2:
			if r&1:
				a[0] = 1
				a[1] ^= 1
		print(" ".join(str(i) for i in a))



